### PR TITLE
[MetricsAdvisor] Added Authentication and CredentiaId properties to Data Feed Sources

### LINKED ISSUE #18609: [MetricsAdvisor] Credential Entity
Add support for new API Credential Entity
- [ ] CRUD operations
- [ ] Create / Update data feed with credential entity

### PR DESCRIPTION
**Pull request by kinelski**
Fixes https://github.com/azure/azure-sdk-for-net/issues/18609.

### CODE REVIEW

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs [hunk 1]
```diff
@@ -455,9 +455,15 @@ public partial class AzureApplicationInsightsDataFeedSource : Azure.AI.MetricsAd
     public partial class AzureBlobDataFeedSource : Azure.AI.MetricsAdvisor.Models.DataFeedSource
     {
         public AzureBlobDataFeedSource(string connectionString, string container, string blobTemplate) { }
+        public Azure.AI.MetricsAdvisor.Models.AzureBlobDataFeedSource.AuthenticationType? Authentication { get { throw null; } set { } }
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs:458]: In general, the flow is:
- User creates a `DatasourceCredential` using a Create operation:
```cs
var credential = new ServicePrincipalDatasourceCredential("clientId", "clientSecret", "tenantId");
credential = adminClient.CreateDatasourceCredential(credential);
string credentialId = credential.Id;
```
- With the credential ID, they may create a data feed with a specific authentication for its data source:
```cs
var dataFeed = new DataFeed()
{
  DataSource = new SqlServerDataFeedSource("connString", "query")
  {
    Authentication = SqlServerDataFeedSource.AuthenticationType.ServicePrincipal,
    DatasourceCredentialId = credentialId
  }
}
dataFeed = adminClient.CreateDataFeed(dataFeed);
```

In particular, the Managed Identity authentication does not require a datasource credential.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeed.cs [hunk 2]
```diff
@@ -198,6 +198,8 @@ internal DataFeedDetail GetDataFeedDetail()
                 detail.Viewers.Add(viewer);
             }
 
+            SetAuthenticationProperties(detail, DataSource);
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeed.cs:201]: We actually have three models: a public one, an internal one used by the service, and another internal one (patch) used by the service during Update.

This method makes the transform public => service.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeed.cs [hunk 3]
```diff
@@ -247,7 +249,53 @@ internal DataFeedDetailPatch GetPatchModel()
             patch.Admins = AdministratorsEmails;
             patch.Viewers = ViewersEmails;
 
+            SetAuthenticationProperties(patch, DataSource);
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeed.cs:252]: This one makes the transform public => patch.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeed.cs [hunk 4]
```diff
@@ -247,7 +249,53 @@ internal DataFeedDetailPatch GetPatchModel()
...
+                    detail.CredentialId = s.DatasourceCredentialId;
+                    break;
+                case AzureDataLakeStorageGen2DataFeedSource s:
+                    detail.AuthenticationType = s.GetAuthenticationTypeEnum();
+                    detail.CredentialId = s.DatasourceCredentialId;
+                    break;
+                case SqlServerDataFeedSource s:
+                    detail.AuthenticationType = s.GetAuthenticationTypeEnum();
+                    detail.CredentialId = s.DatasourceCredentialId;
+                    break;
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeed.cs:275]: These `AuthenticationType` are all different types (`AzureBlobDataFeedSource.AuthenticationType`, `AzureDataExplorerDataFeedSource.AuthenticationType`, etc), while the service uses a single type: `GetAuthenticationTypeEnum` (internal).

We decided to break this enum into multiple enums in .NET to make it easier for users to know which authentication types are valid for each data source. For instance, `AzureBlobDataFeedSource.AuthenticationType` only has `Basic` and `ManagedIdentity`. Users will never make the mistake of trying to set a `ServicePrincipal` authentication in this data source.

According to Ted, auth is one of most confusing things for users starting to get used to the API (as a user myself, I can totally agree), so we decided to split it like this to make it easier for them.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSource.cs [hunk 5]
```diff
@@ -22,15 +22,15 @@ internal DataFeedSource(DataFeedSourceType dataFeedSourceType)
             dataFeedDetail switch
             {
                 AzureApplicationInsightsDataFeed d => new AzureApplicationInsightsDataFeedSource(d.DataSourceParameter),
-                AzureBlobDataFeed d => new AzureBlobDataFeedSource(d.DataSourceParameter),
+                AzureBlobDataFeed d => new AzureBlobDataFeedSource(d.DataSourceParameter, d.AuthenticationType),
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSource.cs:25]: This is the transform service => public. We're now passing the authentication type / credential id returned by the service to some data sources.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/AzureBlobDataFeedSource.cs [hunk 6]
```diff
@@ -58,16 +58,42 @@ public AzureBlobDataFeedSource(string connectionString, string container, string
             BlobTemplate = blobTemplate;
         }
 
-        internal AzureBlobDataFeedSource(AzureBlobParameter parameter)
+        internal AzureBlobDataFeedSource(AzureBlobParameter parameter, AuthenticationTypeEnum? authentication)
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/AzureBlobDataFeedSource.cs:61]: In each data source we're adding a `SetAuthentication` method (transform service => public), and a `GetAuthenticationTypeEnum` method (transform public => service).

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/CodeGenInternal/AuthenticationTypeEnum.cs [hunk 7]
```diff
@@ -3,7 +3,6 @@
 
 namespace Azure.AI.MetricsAdvisor.Models
 {
-    // TODO: expose it as part of 1.0.0-beta.4
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/CodeGenInternal/AuthenticationTypeEnum.cs:6]: Explained in other comments, but we decided to not expose the `AuthenticationTypeEnum` type.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,8 @@
 
 ### New Features
 - Added `DatasourceCredential` CRUD operations to the `MetricsAdvisorAdministrationClient`. This API provides new ways of authenticating to a `DataFeedSource`.
+- Added property `Authentication` to data sources `AzureBlobDataFeedSource`, `AzureDataExplorerDataFeedSource`, `AzureDataLakeStorageGen2DataFeedSource`, and `SqlServerDataFeedSource` to specify the authentication type to use.
```
Comment (maririos) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/CHANGELOG.md:7]: are you thinking of adding some samples to show this and guide the user on the most common data sources and the different authentication types?

**Review by maririos — APPROVED** [review]

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/CHANGELOG.md (hunk 8 above)
Comment (kinelski) [reply sdk/metricsadvisor/Azure.AI.MetricsAdvisor/CHANGELOG.md:7]: Yes. Already tracked by https://github.com/Azure/azure-sdk-for-net/issues/21586.

**Pull request merged**